File: core/Sources/ChotkiCore/Store/SQLiteStore.swift
import Foundation
import CSQLite

private let SQLITE_TRANSIENT = unsafeBitCast(-1, to: sqlite3_destructor_type.self)

/// SQLite persistence over the C API directly.
///
/// Enums with associated values (`Recurrence`) are stored as JSON text. They are
/// never queried on, only read back whole, so a column per case would buy
/// nothing and cost a migration every time a recurrence kind is added.
///
/// Dates are stored as `YYYY-MM-DD` text, which sorts lexicographically — SQLite
/// can range-scan them with no date handling of its own, and the file stays
/// readable by anything.
public final class SQLiteStore: Store, @unchecked Sendable {

    private var db: OpaquePointer?
    private let lock = NSLock()
    private let encoder = JSONEncoder()
    private let decoder = JSONDecoder()

    public init(path: String) throws {
        var handle: OpaquePointer?
        let flags = SQLITE_OPEN_READWRITE | SQLITE_OPEN_CREATE | SQLITE_OPEN_FULLMUTEX
        guard sqlite3_open_v2(path, &handle, flags, nil) == SQLITE_OK, let handle else {
            throw StoreError.open(String(cString: sqlite3_errmsg(handle)))
        }
        self.db = handle
        try exec("PRAGMA journal_mode = WAL;")
        try exec("PRAGMA foreign_keys = ON;")
        try migrate()
    }

    /// A private database that never touches disk. Used by the test suite.
    public static func inMemory() throws -> SQLiteStore {
        try SQLiteStore(path: ":memory:")
    }

    deinit { if let db { sqlite3_close_v2(db) } }

    // MARK: raw helpers

    /// Internal rather than private so tests can build a historical schema and
    /// prove the migration path, which a fresh database never exercises.
    func exec(_ sql: String) throws {
        var error: UnsafeMutablePointer<CChar>?
        guard sqlite3_exec(db, sql, nil, nil, &error) == SQLITE_OK else {
            let message = error.map { String(cString: $0) } ?? "unknown"
            sqlite3_free(error)
            throw StoreError.query("\(message) — while running: \(sql)")
        }
    }

    private func prepare(_ sql: String) throws -> OpaquePointer {
        var statement: OpaquePointer?
        guard sqlite3_prepare_v2(db, sql, -1, &statement, nil) == SQLITE_OK, let statement else {
            throw StoreError.query(String(cString: sqlite3_errmsg(db)))
        }
        return statement
    }

    private func bind(_ statement: OpaquePointer, _ values: [String?]) {
        for (index, value) in values.enumerated() {
            let position = Int32(index + 1)
            if let value {
                sqlite3_bind_text(statement, position, value, -1, SQLITE_TRANSIENT)
            } else {
                sqlite3_bind_null(statement, position)
            }
        }
    }

    private func run(_ sql: String, _ values: [String?]) throws {
        let statement = try prepare(sql)
        defer { sqlite3_finalize(statement) }
        bind(statement, values)
        guard sqlite3_step(statement) == SQLITE_DONE else {
            throw StoreError.query(String(cString: sqlite3_errmsg(db)))
        }
    }

    private func query<T>(_ sql: String, _ values: [String?], _ row: (OpaquePointer) throws -> T?) throws -> [T] {
        let statement = try prepare(sql)
        defer { sqlite3_finalize(statement) }
        bind(statement, values)
        var results: [T] = []
        while sqlite3_step(statement) == SQLITE_ROW {
            if let value = try row(statement) { results.append(value) }
        }
        return results
    }

    private func text(_ statement: OpaquePointer, _ column: Int32) -> String? {
        guard let pointer = sqlite3_column_text(statement, column) else { return nil }
        return String(cString: pointer)
    }

    private func locked<T>(_ body: () throws -> T) rethrows -> T {
        lock.lock(); defer { lock.unlock() }
        return try body()
    }

    // MARK: schema

    private func migrate() throws {
        try exec("CREATE TABLE IF NOT EXISTS schema_version (version INTEGER NOT NULL);")
        let versions: [Int] = try query("SELECT version FROM schema_version;", []) {
            Int(sqlite3_column_int(($0), 0))
        }
        let current = versions.max() ?? 0

        if current < 1 {
            try exec("""
                CREATE TABLE rule (
                    id TEXT PRIMARY KEY,
                    title TEXT NOT NULL,
                    note TEXT,
                    source TEXT,
                    recurrence TEXT NOT NULL,
                    time_of_day TEXT,
                    category TEXT,
                    created_at TEXT NOT NULL,
                    archived_at TEXT
                );

                CREATE TABLE activation (
                    id TEXT PRIMARY KEY,
                    rule_id TEXT NOT NULL REFERENCES rule(id) ON DELETE CASCADE,
                    from_date TEXT NOT NULL,
                    to_date TEXT
                );
                CREATE INDEX activation_by_rule ON activation(rule_id);

                -- One deviation per rule per day. The unique constraint is the
                -- model's rule made structural: a day cannot be both completed
                -- and skipped.
                CREATE TABLE occurrence (
                    id TEXT PRIMARY KEY,
                    rule_id TEXT NOT NULL REFERENCES rule(id) ON DELETE CASCADE,
                    date TEXT NOT NULL,
                    status TEXT NOT NULL,
                    completed_at TEXT,
                    moved_to TEXT,
                    UNIQUE(rule_id, date)
                );
                CREATE INDEX occurrence_by_date ON occurrence(date);

                INSERT INTO schema_version (version) VALUES (1);
                """)
        }

        if current < 2 {
            // Keyed by CIVIL date. The API answers a civil request with the
            // date in the requested reckoning, so keying on what it reports
            // would misfile every Old Calendar day by thirteen days.
            try exec("""
                CREATE TABLE liturgical_day (
                    civil_date TEXT NOT NULL,
                    reckoning TEXT NOT NULL,
                    payload TEXT NOT NULL,
                    fetched_at TEXT NOT NULL,
                    PRIMARY KEY (civil_date, reckoning)
                );
                INSERT INTO schema_version (version) VALUES (2);
                """)
        }

        if current < 3 {
            // Per-rule reminder settings. Nullable, so rules written before this
            // existed keep working and fall back to the default.
            try exec("""
                ALTER TABLE rule ADD COLUMN reminders TEXT;
                INSERT INTO schema_version (version) VALUES (3);
                """)
        }

        if current < 4 {
            // Settings, kept beside the data. They were previously in
            // UserDefaults, where they were not persisting at all — and where
            // they would not have travelled with a backup or to another
            // platform even if they had.
            try exec("""
                CREATE TABLE app_settings (
                    id INTEGER PRIMARY KEY CHECK (id = 1),
                    payload TEXT NOT NULL
                );
                INSERT INTO schema_version (version) VALUES (4);
                """)
        }

        if current < 5 {
            // The prayers a rule carries. Nullable, so rules written before
            // this existed keep working.
            try exec("""
                ALTER TABLE rule ADD COLUMN prayer_ids TEXT;
                INSERT INTO schema_version (version) VALUES (5);
                """)
        }
    }

    // MARK: liturgical cache

    public func saveLiturgicalDay(_ day: LiturgicalDay) throws {
        try locked {
            guard let payload = try encodeJSON(day) else { throw StoreError.query("encode failed") }
            try run("""
                INSERT INTO liturgical_day (civil_date, reckoning, payload, fetched_at)
                VALUES (?, ?, ?, ?)
                ON CONFLICT(civil_date, reckoning) DO UPDATE SET
                    payload = excluded.payload, fetched_at = excluded.fetched_at;
                """, [day.civilDate.iso, day.reckoning.rawValue, payload, encode(day.fetchedAt)])
        }
    }

    public func liturgicalDay(civilDate: CalendarDate, reckoning: Reckoning) throws -> LiturgicalDay? {
        try locked {
            try query(
                "SELECT payload FROM liturgical_day WHERE civil_date = ? AND reckoning = ?;",
                [civilDate.iso, reckoning.rawValue]
            ) { try decodeJSON(LiturgicalDay.self, text($0, 0)) }.first
        }
    }

    public func liturgicalDays(
        reckoning: Reckoning, from: CalendarDate, through: CalendarDate
    ) throws -> [LiturgicalDay] {
        try locked {
            try query("""
                SELECT payload FROM liturgical_day
                WHERE reckoning = ? AND civil_date >= ? AND civil_date <= ?
                ORDER BY civil_date;
                """, [reckoning.rawValue, from.iso, through.iso]) {
                try decodeJSON(LiturgicalDay.self, text($0, 0))
            }
        }
    }

    public func clearLiturgicalCache(reckoning: Reckoning?) throws {
        try locked {
            if let reckoning {
                try run("DELETE FROM liturgical_day WHERE reckoning = ?;", [reckoning.rawValue])
            } else {
                try run("DELETE FROM liturgical_day;", [])
            }
        }
    }

    // MARK: coding helpers

    private func encodeJSON<T: Encodable>(_ value: T?) throws -> String? {
        guard let value else { return nil }
        return String(data: try encoder.encode(value), encoding: .utf8)
    }

    private func decodeJSON<T: Decodable>(_ type: T.Type, _ string: String?) throws -> T? {
        guard let string, let data = string.data(using: .utf8) else { return nil }
        return try decoder.decode(type, from: data)
    }

    /// An instance rather than a shared static: every call site runs inside
    /// `locked`, so the formatter is never touched from two threads at once.
    private let stamp: ISO8601DateFormatter = {
        let f = ISO8601DateFormatter()
        f.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        return f
    }()

    private func encode(_ date: Date?) -> String? { date.map { stamp.string(from: $0) } }
    private func decode(_ string: String?) -> Date? { string.flatMap { stamp.date(from: $0) } }

    // MARK: rules

    public func save(_ rule: Rule) throws {
        try locked {
            try run("""
                INSERT INTO rule (id, title, note, source, recurrence, time_of_day, category, created_at, archived_at, reminders, prayer_ids)
                VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
                ON CONFLICT(id) DO UPDATE SET
                    title = excluded.title, note = excluded.note, source = excluded.source,
                    recurrence = excluded.recurrence, time_of_day = excluded.time_of_day,
                    category = excluded.category, archived_at = excluded.archived_at,
                    reminders = excluded.reminders, prayer_ids = excluded.prayer_ids;
                """, [
                    rule.id.uuidString, rule.title, rule.note, rule.source,
                    try encodeJSON(rule.recurrence), try encodeJSON(rule.timeOfDay),
                    rule.category, encode(rule.createdAt), encode(rule.archivedAt),
                    try encodeJSON(rule.reminders), try encodeJSON(rule.prayerIDs)
                ])
        }
    }

    private func decodeRule(_ s: OpaquePointer) throws -> Rule? {
        guard let id = text(s, 0).flatMap(UUID.init(uuidString:)),
              let title = text(s, 1),
              let recurrence = try decodeJSON(Recurrence.self, text(s, 4)),
              let createdAt = decode(text(s, 7))
        else { return nil }
        return Rule(
            id: id, title: title, note: text(s, 2), source: text(s, 3),
            recurrence: recurrence,
            timeOfDay: try decodeJSON(TimeOfDay.self, text(s, 5)),
            category: text(s, 6),
            reminders: try decodeJSON(RuleReminders.self, text(s, 9)),
            prayerIDs: try decodeJSON([String].self, text(s, 10)),
            createdAt: createdAt, archivedAt: decode(text(s, 8))
        )
    }

    private static let ruleColumns =
        "id, title, note, source, recurrence, time_of_day, category, created_at, archived_at, reminders, prayer_ids"

    public func rule(id: UUID) throws -> Rule? {
        try locked {
            try query("SELECT \(Self.ruleColumns) FROM rule WHERE id = ?;", [id.uuidString]) {
                try decodeRule($0)
            }.first
        }
    }

    public func rules(includeArchived: Bool) throws -> [Rule] {
        let filter = includeArchived ? "" : "WHERE archived_at IS NULL"
        return try locked {
            try query("SELECT \(Self.ruleColumns) FROM rule \(filter) ORDER BY created_at;", []) {
                try decodeRule($0)
            }
        }
    }

    // MARK: activations

    public func save(_ activation: Activation) throws {
        try locked {
            try run("""
                INSERT INTO activation (id, rule_id, from_date, to_date) VALUES (?, ?, ?, ?)
                ON CONFLICT(id) DO UPDATE SET from_date = excluded.from_date, to_date = excluded.to_date;
                """, [
                    activation.id.uuidString, activation.ruleID.uuidString,
                    activation.from.iso, activation.to?.iso
                ])
        }
    }

    public func removeActivation(id: UUID) throws {
        try locked { try run("DELETE FROM activation WHERE id = ?;", [id.uuidString]) }
    }

    public func activations(ruleID: UUID?) throws -> [Activation] {
        let filter = ruleID == nil ? "" : "WHERE rule_id = ?"
        let values = ruleID.map { [$0.uuidString] } ?? []
        return try locked {
            try query("SELECT id, rule_id, from_date, to_date FROM activation \(filter) ORDER BY from_date;", values) { s in
                guard let id = text(s, 0).flatMap(UUID.init(uuidString:)),
                      let ruleID = text(s, 1).flatMap(UUID.init(uuidString:)),
                      let from = text(s, 2).flatMap(CalendarDate.init(iso:))
                else { return nil }
                return Activation(id: id, ruleID: ruleID, from: from,
                                  to: text(s, 3).flatMap(CalendarDate.init(iso:)))
            }
        }
    }

    // MARK: occurrences

    public func save(_ occurrence: Occurrence) throws {
        try locked {
            try run("""
                INSERT INTO occurrence (id, rule_id, date, status, completed_at, moved_to)
                VALUES (?, ?, ?, ?, ?, ?)
                ON CONFLICT(rule_id, date) DO UPDATE SET
                    status = excluded.status, completed_at = excluded.completed_at,
                    moved_to = excluded.moved_to;
                """, [
                    occurrence.id.uuidString, occurrence.ruleID.uuidString, occurrence.date.iso,
                    occurrence.status.rawValue, encode(occurrence.completedAt), occurrence.movedTo?.iso
                ])
        }
    }

    public func occurrences(
        ruleID: UUID?, from: CalendarDate?, through: CalendarDate?
    ) throws -> [Occurrence] {
        var clauses: [String] = []
        var values: [String?] = []
        if let ruleID { clauses.append("rule_id = ?"); values.append(ruleID.uuidString) }
        if let from { clauses.append("date >= ?"); values.append(from.iso) }
        if let through { clauses.append("date <= ?"); values.append(through.iso) }
        let filter = clauses.isEmpty ? "" : "WHERE " + clauses.joined(separator: " AND ")

        return try locked {
            try query("""
                SELECT id, rule_id, date, status, completed_at, moved_to
                FROM occurrence \(filter) ORDER BY date, rule_id;
                """, values) { s in
                guard let id = text(s, 0).flatMap(UUID.init(uuidString:)),
                      let ruleID = text(s, 1).flatMap(UUID.init(uuidString:)),
                      let date = text(s, 2).flatMap(CalendarDate.init(iso:)),
                      let status = text(s, 3).flatMap(OccurrenceStatus.init(rawValue:))
                else { return nil }
                return Occurrence(id: id, ruleID: ruleID, date: date, status: status,
                                  completedAt: decode(text(s, 4)),
                                  movedTo: text(s, 5).flatMap(CalendarDate.init(iso:)))
            }
        }
    }

    public func removeOccurrence(ruleID: UUID, date: CalendarDate) throws {
        try locked {
            try run("DELETE FROM occurrence WHERE rule_id = ? AND date = ?;",
                    [ruleID.uuidString, date.iso])
        }
    }

    // MARK: settings

    public func loadSettings() throws -> AppSettings? {
        try locked {
            let rows: [AppSettings] = try query("SELECT payload FROM app_settings WHERE id = 1;", []) { s in
                try decodeJSON(AppSettings.self, text(s, 0))
            }
            return rows.first
        }
    }

    public func saveSettings(_ settings: AppSettings) throws {
        try locked {
            try run("""
                INSERT INTO app_settings (id, payload) VALUES (1, ?)
                ON CONFLICT(id) DO UPDATE SET payload = excluded.payload;
                """, [try encodeJSON(settings)])
        }
    }

    // MARK: atomic edits

    /// A split that closed the old stretch but failed to open the new one would
    /// make a rule silently disappear. All of it lands, or none of it does.
    public func apply(_ plan: EditPlan) throws {
        try locked { try exec("BEGIN IMMEDIATE;") }
        do {
            for rule in plan.updatedRules + plan.newRules { try save(rule) }
            for activation in plan.updatedActivations + plan.newActivations { try save(activation) }
            for id in plan.removedActivationIDs { try removeActivation(id: id) }
            for occurrence in plan.newOccurrences { try save(occurrence) }
            try locked { try exec("COMMIT;") }
        } catch {
            try? locked { try exec("ROLLBACK;") }
            throw error
        }
    }
}

File: android/core/src/main/kotlin/org/chotki/core/store/SqliteStore.kt
package org.chotki.core.store

import kotlinx.serialization.json.Json
import org.chotki.core.Activation
import org.chotki.core.CalendarDate
import org.chotki.core.AppSettings
import org.chotki.core.EditPlan
import org.chotki.core.LiturgicalDay
import org.chotki.core.Occurrence
import org.chotki.core.Reckoning
import org.chotki.core.OccurrenceStatus
import org.chotki.core.Recurrence
import org.chotki.core.Rule
import org.chotki.core.RuleCategory
import org.chotki.core.TimeOfDay
import java.time.Instant
import java.util.UUID

/**
 * The store, written once against [Db] and run on whatever supplies one.
 *
 * Dates are ISO text so SQLite can order and range-scan them with no date
 * handling of its own. Structured fields are JSON.
 */
class SqliteStore(private val db: Db) : Store {

    init {
        db.execute("PRAGMA journal_mode=WAL;")
        db.execute("PRAGMA foreign_keys=ON;")
        Schema.migrate(db)
    }

    private val json = Json { encodeDefaults = true; ignoreUnknownKeys = true }

    // MARK: rules

    override fun save(rule: Rule) {
        db.update(
            """
            INSERT INTO rule (${Schema.RULE_COLUMNS})
            VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
            ON CONFLICT(id) DO UPDATE SET
                title = excluded.title, note = excluded.note, source = excluded.source,
                recurrence = excluded.recurrence, time_of_day = excluded.time_of_day,
                category = excluded.category, archived_at = excluded.archived_at,
                reminders = excluded.reminders, prayer_ids = excluded.prayer_ids,
                hidden_from_library = excluded.hidden_from_library;
            """.trimIndent(),
            listOf(
                rule.id.toString(),
                rule.title,
                rule.note,
                rule.source,
                json.encodeToString(Recurrence.serializer(), rule.recurrence),
                rule.timeOfDay?.let { json.encodeToString(TimeOfDay.serializer(), it) },
                rule.category?.name,
                rule.createdAt.toString(),
                rule.archivedAt?.toString(),
                // Reminders arrive at phase 7. The column is written null rather
                // than dropped, so the row shape stays right and nothing has to
                // migrate again when they do.
                null,
                rule.prayerIDs?.let { json.encodeToString(it) },
                // Only when set, so absent goes on meaning "still offered".
                if (rule.hiddenFromLibrary == true) "1" else null,
            ),
        )
    }

    override fun rule(id: UUID): Rule? =
        db.query(
            "SELECT ${Schema.RULE_COLUMNS} FROM rule WHERE id = ?;",
            listOf(id.toString()),
        ) { readRule(it) }.firstOrNull()

    override fun rules(includeArchived: Boolean): List<Rule> {
        val where = if (includeArchived) "" else "WHERE archived_at IS NULL "
        return db.query(
            "SELECT ${Schema.RULE_COLUMNS} FROM rule ${where}ORDER BY created_at;",
        ) { readRule(it) }
    }

    private fun readRule(row: Row): Rule {
        val id = row.string(0) ?: throw StoreException("a rule with no id")
        val recurrence = row.string(4) ?: throw StoreException("a rule with no recurrence")
        return Rule(
            id = UUID.fromString(id),
            title = row.string(1) ?: throw StoreException("a rule with no title"),
            note = row.string(2),
            source = row.string(3),
            recurrence = json.decodeFromString(Recurrence.serializer(), recurrence),
            timeOfDay = row.string(5)?.let { json.decodeFromString(TimeOfDay.serializer(), it) },
            category = row.string(6)?.let { name ->
                RuleCategory.entries.firstOrNull { it.name == name }
            },
            prayerIDs = row.string(10)?.let { json.decodeFromString<List<String>>(it) },
            createdAt = Instant.parse(
                row.string(7) ?: throw StoreException("a rule with no created_at"),
            ),
            archivedAt = row.string(8)?.let(Instant::parse),
            hiddenFromLibrary = row.int(11)?.let { it == 1 },
        )
    }

    // MARK: activations

    override fun save(activation: Activation) {
        db.update(
            """
            INSERT INTO activation (id, rule_id, from_date, to_date) VALUES (?, ?, ?, ?)
            ON CONFLICT(id) DO UPDATE SET
                rule_id = excluded.rule_id, from_date = excluded.from_date,
                to_date = excluded.to_date;
            """.trimIndent(),
            listOf(
                activation.id.toString(),
                activation.ruleID.toString(),
                activation.from.iso,
                activation.to?.iso,
            ),
        )
    }

    override fun removeActivation(id: UUID) {
        db.update("DELETE FROM activation WHERE id = ?;", listOf(id.toString()))
    }

    override fun activations(ruleID: UUID?): List<Activation> {
        val where = if (ruleID == null) "" else "WHERE rule_id = ? "
        val args = listOfNotNull(ruleID?.toString())
        return db.query(
            "SELECT id, rule_id, from_date, to_date FROM activation ${where}ORDER BY from_date;",
            args,
        ) { row ->
            Activation(
                id = UUID.fromString(row.string(0)),
                ruleID = UUID.fromString(row.string(1)),
                from = CalendarDate.parse(row.string(2)!!)
                    ?: throw StoreException("an activation with no start"),
                to = row.string(3)?.let(CalendarDate::parse),
            )
        }
    }

    // MARK: occurrences

    override fun save(occurrence: Occurrence) {
        db.update(
            """
            INSERT INTO occurrence (id, rule_id, date, status, completed_at, moved_to)
            VALUES (?, ?, ?, ?, ?, ?)
            ON CONFLICT(rule_id, date) DO UPDATE SET
                status = excluded.status, completed_at = excluded.completed_at,
                moved_to = excluded.moved_to;
            """.trimIndent(),
            listOf(
                occurrence.id.toString(),
                occurrence.ruleID.toString(),
                occurrence.date.iso,
                occurrence.status.name,
                occurrence.completedAt?.toString(),
                occurrence.movedTo?.iso,
            ),
        )
    }

    override fun occurrences(
        ruleID: UUID?,
        from: CalendarDate?,
        through: CalendarDate?,
    ): List<Occurrence> {
        val clauses = mutableListOf<String>()
        val args = mutableListOf<String?>()
        ruleID?.let { clauses.add("rule_id = ?"); args.add(it.toString()) }
        // ISO dates compare correctly as text, which is the whole reason for
        // storing them that way.
        from?.let { clauses.add("date >= ?"); args.add(it.iso) }
        through?.let { clauses.add("date <= ?"); args.add(it.iso) }
        val where = if (clauses.isEmpty()) "" else "WHERE ${clauses.joinToString(" AND ")} "

        return db.query(
            """
            SELECT id, rule_id, date, status, completed_at, moved_to
            FROM occurrence ${where}ORDER BY date;
            """.trimIndent(),
            args,
        ) { row ->
            Occurrence(
                id = UUID.fromString(row.string(0)),
                ruleID = UUID.fromString(row.string(1)),
                date = CalendarDate.parse(row.string(2)!!)
                    ?: throw StoreException("an occurrence with no date"),
                status = OccurrenceStatus.valueOf(row.string(3)!!),
                completedAt = row.string(4)?.let(Instant::parse),
                movedTo = row.string(5)?.let(CalendarDate::parse),
            )
        }
    }

    override fun removeOccurrence(ruleID: UUID, date: CalendarDate) {
        db.update(
            "DELETE FROM occurrence WHERE rule_id = ? AND date = ?;",
            listOf(ruleID.toString(), date.iso),
        )
    }

    // MARK: the church calendar

    override fun saveLiturgicalDay(day: LiturgicalDay) {
        db.update(
            """
            INSERT INTO liturgical_day (civil_date, reckoning, payload, fetched_at)
            VALUES (?, ?, ?, ?)
            ON CONFLICT(civil_date, reckoning) DO UPDATE SET
                payload = excluded.payload, fetched_at = excluded.fetched_at;
            """.trimIndent(),
            listOf(
                day.civilDate.iso,
                day.reckoning.name,
                json.encodeToString(LiturgicalDay.serializer(), day),
                day.fetchedAt.toString(),
            ),
        )
    }

    override fun liturgicalDay(civilDate: CalendarDate, reckoning: Reckoning): LiturgicalDay? =
        db.query(
            "SELECT payload FROM liturgical_day WHERE civil_date = ? AND reckoning = ?;",
            listOf(civilDate.iso, reckoning.name),
        ) { readDay(it) }.firstOrNull()

    override fun liturgicalDays(
        reckoning: Reckoning,
        from: CalendarDate,
        through: CalendarDate,
    ): List<LiturgicalDay> = db.query(
        """
        SELECT payload FROM liturgical_day
        WHERE reckoning = ? AND civil_date >= ? AND civil_date <= ?
        ORDER BY civil_date;
        """.trimIndent(),
        listOf(reckoning.name, from.iso, through.iso),
    ) { readDay(it) }

    private fun readDay(row: Row): LiturgicalDay =
        json.decodeFromString(
            LiturgicalDay.serializer(),
            row.string(0) ?: throw StoreException("a cached day with no payload"),
        )

    override fun clearLiturgicalCache(reckoning: Reckoning?) {
        if (reckoning == null) {
            db.update("DELETE FROM liturgical_day;")
        } else {
            db.update("DELETE FROM liturgical_day WHERE reckoning = ?;", listOf(reckoning.name))
        }
    }

    // MARK: settings

    override fun loadSettings(): AppSettings? =
        db.query("SELECT payload FROM app_settings WHERE id = 1;") { row ->
            json.decodeFromString(AppSettings.serializer(), row.string(0)!!)
        }.firstOrNull()

    override fun saveSettings(settings: AppSettings) {
        db.update(
            """
            INSERT INTO app_settings (id, payload) VALUES (1, ?)
            ON CONFLICT(id) DO UPDATE SET payload = excluded.payload;
            """.trimIndent(),
            listOf(json.encodeToString(AppSettings.serializer(), settings)),
        )
    }

    // MARK: the whole plan, or none of it

    override fun apply(plan: EditPlan) {
        db.transaction {
            for (rule in plan.updatedRules + plan.newRules) save(rule)
            for (activation in plan.updatedActivations + plan.newActivations) save(activation)
            for (id in plan.removedActivationIDs) removeActivation(id)
            for (occurrence in plan.newOccurrences) save(occurrence)
        }
    }

    override fun close() = db.close()
}
